### ISSUE TITLE
Subpixel Precision

### ISSUE THREAD

**Issue by ADormant** [devmiyax/yabause#193]
Both Saturn and PS1 lack subpixel precision and perspective-correct texture mapping which cause shaking polygons however PS1 emulators have ''GTE Accuracy'' option mitigating this problem. Is something like this possible for Saturn emulation? @devmiyax 

https://github.com/simias/rustation/tree/subpixel
https://github.com/tapcio/PeteOpenGL2Tweak/tree/master/gpuPeteOpenGL2Tweak
https://github.com/simias/rustation/commit/4c6933604d23aed3e874ad58427d8fd3b563ee9e
https://github.com/simias/rustation/commit/aea0850102b32c8e3ccd86cf434d1f0e7666e2c9
https://github.com/simias/rustation/commit/7d10ed3c548bd2133d2e7fc5b5efd777c3782b48
https://github.com/simias/rustation/commit/14195bc46802c91248de802ceaa967a2f7d8fa3b
https://github.com/simias/rustation/commit/10d1ae01c1a651a8949539cd283f448e24e4940c
https://github.com/libretro/beetle-psx-libretro/commit/9cc93ebde6a866d82c8d4d00019f52c810ebe597
https://github.com/simias/beetle-psx-libretro/commit/2b5c6f623affe5fffa1ddcb6894dbaf25d91d43d
https://github.com/simias/rustation/commit/4b0b5f28b01c3cc56c8f5789b06fcd276153dbc5
https://pcsxr.codeplex.com/discussions/264234

**Comment by devmiyax** [comment]
I remember that designers tried to draw  perspective-corrected texture ｍanually for PS1.
Actually Saturn has perspective correction on VDP1. Conversely I need to emulate perspective correction without z value. 

https://www.youtube.com/watch?v=TzITZgtchMU

That's problem for Saturn emulation. Still this function has bug. sigh.
https://github.com/devmiyax/yabause/issues/184

**Comment by ADormant** [comment]
@devmiyax Did it have real perspective-correct texture mapping or just perspective correction thanks to quads like this PS1 modification 
https://github.com/simias/rustation/commit/4b0b5f28b01c3cc56c8f5789b06fcd276153dbc5
Everywhere on the net says it had affine texture mapping,. Saturn didn't have Depth Buffer either though it may be possible to implement perspective correct texture mapping even without it. It was theorized that even Z buffer may be possible to do for PS1(perhaps for Saturn emulation too) emulation though that would be the hardest to do.
You can see lack of subpixel precision and probably perspective correction in this trailer notice jittering walls. https://www.youtube.com/watch?v=kUe9ASlu9Us
Basic subpixel precision would be easiest to implement if at all possible for Saturn since it's the simplest rendering enhancement for PS1.

PS1 emulation experiments perspective correct mapping
https://www.youtube.com/watch?v=inFqJvEGGYc
Z Buffer+perspective correct mapping
https://www.youtube.com/watch?v=zGnLRR9kyvc

https://www.youtube.com/watch?v=Jglm4Jkkq5I

https://www.youtube.com/watch?v=brC96Rtmp2Y

Draw distance hack
https://www.youtube.com/watch?v=CRBcvSrWr4k

**Comment by ADormant** [comment]
Subpixel precision
https://www.youtube.com/watch?v=uSbU5B2SnRM

**Comment by devmiyax** [comment]
It seems that I didn't get what you want to say.
Saturn perspective correction is not the real perspective correction, just thanks for quads based rendering.

In my understanding 'GTE Accuracy' realize vertex calculation accuracy modify the input to GTE. Right?

Saturn doesn't have GTE like module. All vertex calculations are done on CPU or SCU.
Since these chips are programmable, Software patch to each game is need to do same thing as 'GTE Accuracy'. I think it's impossible.

There is one exception. RBG( Rotate Background Screen ) accepts matrix and transform information and calculate each coordinate pixel using them. HD RBG might be possible.

**Comment by ADormant** [comment]
These processors being programmable indeed seem to make it harder in Saturn's case. What do you mean by HD RBG?
Though in case of more advanced GTE Accuracy some modifications to the emulatated CPU registers also seem necessary perhaps @simias can explain better how subpixel precision and perspective correct texture mapping work in PS1 emulators.


https://ia801705.us.archive.org/20/items/SCU_Users_Manual_Third_Version_1994_Sega/SCU_Users_Manual_Third_Version_1994_Sega.pdf

**Comment by simias** [comment]
I don't know much about Saturn but from your description it looks like it's worse than on the PSX. The GTE, while a CPU coprocessor, has a well defined "high level" interface and you don't really have to do a lot of guesswork to figure out what's happening.

You got a bunch of registers where you can put your 3D vertex coordinates, elsewhere you put your transformation matrices and you have a bunch of opcodes to transform and project your coordinates. For subpixel precision hacks you just do those calculations with increased precision and use those values instead of the "native" ones when you draw the triangles in the GPU. Pretty straightforward.

If the Saturn doesn't have any kind of hardware accelerated 3D transformation module and everything is done in software on the CPU then I can't really imagine how you could make a generic hack. Maybe if you're lucky the console's BIOS provides optimized functions to do those transformations and you could hook those up HLE-style?

**Comment by ADormant** [comment]
By the way @simias what is needed for perspective correct mapping hacks? I remember there was something about passing z coordinates between GTE and GPU.

**Comment by simias** [comment]
Yeah it's related. On the real console the Z coordinate is "lost" in the GTE after the vertex is projected on the screen since the GPU doesn't know how to use it anyway.

Perspective correct hacks store the Z coordinates (in the same way you store the extended-precision coordinates for subpixel accuracy) so that the emulated GPU can do the perspective correction.

So generally if you can do subpixel precision you can do perspective correct mapping using the same infrastructure. In theory, at least.

**Comment by devmiyax** [comment]
@simias, Thank you for your information. 'GTE Accuracy' is great idea.

@ADormant, Could you remove the link to the SCU document, because confidential documents should not be on public.

About HD RGB.

You can see 'Rotate BackGround Screen' on many Saturn games like sky and sea of this screenshot.
![_storage_emulated_0_yabause_screenshots_mk-81009_2016_01_07_06_43_51](https://cloud.githubusercontent.com/assets/1174152/13717360/4a1738cc-e824-11e5-96ca-ae07a0a854ff.png)

These pixel information is calculated in VDP2 using high level input like viewpoint, matrix, etc.
https://github.com/devmiyax/yabause/blob/extui/yabause/src/vidshared.c#L195

These calculated pixel information is rounded off to Saturn original resolution(350x220) even if your display is HD(1920x1080).
https://github.com/devmiyax/yabause/blob/extui/yabause/src/vidogl.c#L3106

If VDP2 emulation calculate targeting high reslution or using float value, you will see hi-resolution sea and sky. That's my idea.

**Comment by ADormant** [comment]
I see. Do you think it's feasible to hook up some bios functions related to vertex tranaformations just curious.

**Comment by ADormant** [comment]
@devmiyax Working subpixel precision and perspective correction.
https://drive.google.com/file/d/0Bz8IYcLfu84zSUlVOEJFWGVwenM/view
https://github.com/iCatButler?tab=repositories
